### PR TITLE
C++ API parity: Hardtanh

### PR DESCRIPTION
**Pull request by pbelevich**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* #27382 C++ API parity: Linear
* #27060 C++ API parity: LogSigmoid
* #27059 C++ API parity: LeakyReLU
* **#27038 C++ API parity: Hardtanh**

Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 1]
```diff
@@ -20,6 +20,15 @@ inline Tensor hardshrink(const Tensor& input,
   return torch::hardshrink(input, options.lambda());
 }
 
+inline Tensor hardtanh(const Tensor& input, const HardtanhOptions& options) {
+  if (options.inplace()) {
+    return torch::hardtanh_(const_cast<Tensor&>(input),
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: same here: we might want to just change `input`'s type to `Tensor&`, and remove `const_cast`.

File: torch/csrc/api/src/nn/modules/activation.cpp [hunk 2]
```diff
@@ -36,5 +36,23 @@ void HardshrinkImpl::pretty_print(std::ostream& stream) const {
...
+  return F::hardtanh(input, options);
+}
+
+void HardtanhImpl::reset() {}
+
+void HardtanhImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::Hardtanh(min_val=" << options.min_val()
+         << ", max_val=" << options.max_val()
+         << ", inplace=" << options.inplace() << ")";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/activation.cpp]: same here: whether to print `inplace` is conditional on whether it is `true`, in order to match the Python version.

File: torch/csrc/api/include/torch/nn/options/activation.h [hunk 3]
```diff
@@ -26,5 +26,21 @@ struct HardshrinkOptions {
...
+struct HardtanhOptions {
+  HardtanhOptions() {}
+
+  // minimum value of the linear region range. Default: -1
+  TORCH_ARG(double, min_val) = -1.0;
+
+  // maximum value of the linear region range. Default: 1
+  TORCH_ARG(double, max_val) = 1.0;
+
+  // can optionally do the operation in-place. Default: False
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/activation.h]: nit: need `///` instead of `//` (similarly for other PRs)

**Review by yf225 — APPROVED** [review]
Thanks @pbelevich !

File: torch/csrc/api/src/nn/modules/activation.cpp [hunk 4]
```diff
@@ -38,5 +38,26 @@ void HardshrinkImpl::pretty_print(std::ostream& stream) const {
          << "torch::nn::Hardshrink(" << options.lambda() << ")";
 }
 
+// ============================================================================
+
+HardtanhImpl::HardtanhImpl(const HardtanhOptions& options_)
+    : options(options_) {}
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/activation.cpp]: oops sorry I think we are missing one assert here:
```python
assert self.max_val > self.min_val  # Python version
```

**Review by yf225 — CHANGES_REQUESTED** [review]

File: torch/csrc/api/src/nn/modules/activation.cpp [hunk 5]
```diff
@@ -38,5 +38,29 @@ void HardshrinkImpl::pretty_print(std::ostream& stream) const {
          << "torch::nn::Hardshrink(" << options.lambda() << ")";
 }
 
+// ============================================================================
+
+HardtanhImpl::HardtanhImpl(const HardtanhOptions& options_)
+    : options(options_) {
+  TORCH_CHECK(options.max_val() > options.min_val(),
+              "max_val must be greater than min_val");
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/activation.cpp]: Copying from our discussion: I think ideally we should put the check in module reset(), with the expectation that in the future we will support the use case of changing a module’s options and then call reset() to reinitialize the module (without calling its constructor)

Comment (yf225) [reply torch/csrc/api/src/nn/modules/activation.cpp]: We need to call reset() from the constructor, if we make that change :)

Comment (pbelevich) [reply torch/csrc/api/src/nn/modules/activation.cpp]: fixed

**Review by yf225 — APPROVED** [review]
Thanks a lot @pbelevich !

**Pull request closed without merging**